File: backend/handlers/categories.go
package handlers

import (
	"log"
	"net/http"

	"github.com/go-chi/chi/v5"

	"pcmaxing/backend/db"
)

// CategoriesRouter returns a chi router for /api/categories.
func CategoriesRouter() http.Handler {
	r := chi.NewRouter()
	r.Get("/", listCategories)
	r.Get("/{category}/components", getCategoryComponents)
	return r
}

// GET /api/categories
func listCategories(w http.ResponseWriter, r *http.Request) {
	categories := []string{}
	if err := db.DB.Model(&Component{}).Distinct("category").Pluck("category", &categories).Error; err != nil {
		log.Printf("listCategories error: %v", err)
		writeError(w, http.StatusInternalServerError, "Failed to fetch categories")
		return
	}
	writeJSON(w, http.StatusOK, categories)
}

// GET /api/categories/:category/components
func getCategoryComponents(w http.ResponseWriter, r *http.Request) {
	category := chi.URLParam(r, "category")
	if category == "" || len(category) > maxParamLen {
		writeError(w, http.StatusBadRequest, "Invalid category")
		return
	}

	components := []Component{}
	if err := db.DB.Where("category = ?", category).Order("price ASC").Find(&components).Error; err != nil {
		log.Printf("getCategoryComponents error: %v", err)
		writeError(w, http.StatusInternalServerError, "Failed to fetch components")
		return
	}

	writeJSON(w, http.StatusOK, components)
}

File: backend/routes/categories.js
const { Router } = require('express');
const { supabase } = require('../lib/supabase');

const router = Router();

// GET /api/categories
router.get('/', async (req, res, next) => {
  try {
    const { data, error } = await supabase
      .from('components')
      .select('category')
      .order('category');

    if (error) throw error;
    const categories = [...new Set(data.map((c) => c.category))];
    res.json(categories);
  } catch (err) {
    next(err);
  }
});

// GET /api/categories/:category/components
router.get('/:category/components', async (req, res, next) => {
  try {
    const { data, error } = await supabase
      .from('components')
      .select('*')
      .eq('category', req.params.category)
      .order('price');

    if (error) throw error;
    res.json(data);
  } catch (err) {
    next(err);
  }
});

module.exports = router;
